include/git2/userbuf.h

```diff
@@ -0,0 +1,118 @@
+#define GIT_USERBUF_INIT { NULL, 0, 0 }
+
+/** A constant initialization macro for `git_userbuf` objects. */
+#define GIT_USERBUF_CONST(str, len) { (char *)(str), 0, (size_t)(len) }
+
+/**
+ * A data buffer for exporting data from libgit2.
+ *
+ * Sometimes libgit2 wants to return an allocated data buffer to the
+ * caller and allow the caller take responsibility for its lifecycle.
+ * This requires the caller to free the memory with `git_userbuf_dispose`
+ * when they have finished using it.
+ */
+typedef struct {
+	/**
+	 * The buffer's contents.  This is a NUL terminated C string.
+	 */
+	char *ptr;
+
+	/**
+	 * This is the allocated size of the allocated buffer.  For
+	 * buffers returned to you from libgit2, you should not
+	 * modify this value.  For any buffer that you pass to
+	 * libgit2, this remain 0.
+	 */
+	size_t asize;
+
+	/**
+	 * The size (in bytes) of the data in the buffer, not including
+	 * the NUL terminating character.
+	 */
+	size_t size;
+} git_userbuf;
+
+/**
+ * Check quickly if buffer looks like it contains binary data
+ *
+ * @param buf Buffer to check
+ * @return 1 if buffer looks like non-text data
+ */
+GIT_EXTERN(int) git_userbuf_is_binary(const git_userbuf *buf);
+
+ /**
+ * Check quickly if buffer contains a NUL byte
+ *
+ * @param buf Buffer to check
+ * @return 1 if buffer contains a NUL byte
+ */
+GIT_EXTERN(int) git_userbuf_contains_nul(const git_userbuf *buf);
+
+/**
+ * Places the given data in the buffer.  This is necessary for some
+ * callback functions that take user data.  If there is already data in
+ * the buffer, you should call `git_userbuf_dispose` before setting the
+ * buffer data.
+ *
+ * @param buf The buffer to place data into
+ * @param ptr The data to place in the buffer
+ * @param len The length of the data in bytes
+ */
+GIT_EXTERN(int) git_userbuf_set(git_userbuf *buf, const void *ptr, size_t len);
+
+/**
+ * Resize the buffer allocation to make more space.
+ *
+ * This will attempt to grow the buffer to accommodate the target size.
+ *
+ * If the buffer refers to memory that was not allocated by libgit2 (i.e.
+ * the `asize` field is zero), then `ptr` will be replaced with a newly
+ * allocated block of data.  Be careful so that memory allocated by the
+ * caller is not lost.  As a special variant, if you pass `target_size` as
+ * 0 and the memory is not allocated by libgit2, this will allocate a new
+ * buffer of size `size` and copy the external data into it.
+ *
+ * Currently, this will never shrink a buffer, only expand it.
+ *
+ * If the allocation fails, this will return an error and the buffer will be
+ * marked as invalid for future operations, invaliding the contents.
+ *
+ * @param buffer The buffer to be resized; may or may not be allocated yet
+ * @param target_size The desired available size
+ * @return 0 on success, -1 on allocation failure
+ */
+GIT_EXTERN(int) git_userbuf_grow(git_userbuf *buffer, size_t target_size);
```

Comment:
So if we're going to introduce this type, to me it should really be limited to providing output _to_ the user. As you said, separation of concerns makes sense, and we're forced to handle character arrays to the user in wrapper as opposed to handing memory ownership over to the caller directly. Otherwise there might be mismatches in malloc/free implementations etc.

So the scope of `git_userbuf` should be getting information to the user so that he may access it and then properly dispose of it by calling `git_userbuf_dispose`. To me, `git_userbuf_grow` and `git_userbuf_set` don't fit into this scope and should not be provided.

What's missing though are `git_userbuf_len` and `git_userbuf_ptr` functions to access the structure, mostly because I'd prefer the structure to be opaque to the user. I don't think we'll ever be able to actually make it opaque without a major backwards compatibility breakage (which I definitely don't want to pursue), but at least allowing users to treat the structure as opaque would make sense to me. We should probably also provide `git_userbuf_init`, as the macro won't work in all situations.